logging/src/main.ts

```diff
@@ -1,9 +1,21 @@
 import express, { Express, NextFunction, Request, Response } from "express";
 import { request, ClientRequest, RequestOptions, IncomingMessage } from "http";
+import {writeStatusToDisk} from "./status";
 
 const app: Express = express();
 const port: number = 80;
 
+const serviceLaunchTime = Math.round(new Date().getTime() / 1000);
+const statusFilePath = process.env.STATUS_FILE_PATH || "/opt/orbs/status/status.json";
+
+let error = '';
```

Comment:
I asked ChatGPT for advice about how we can avoid this "global" error variable and it suggested this:

```typescript
import express, { Express, NextFunction, Request, Response } from "express";
import { request, ClientRequest, RequestOptions, IncomingMessage } from "http";
import { writeStatusToDisk } from "./status";

const app: Express = express();
const port: number = 80;

const serviceLaunchTime = Math.round(new Date().getTime() / 1000);
const statusFilePath = process.env.STATUS_FILE_PATH || "/opt/orbs/status/status.json";

setInterval(() => {
    writeStatusToDisk(statusFilePath, serviceLaunchTime);
}, 5 * 60 * 1000);

const validNameRegex = /^[a-zA-Z0-9][a-zA-Z0-9_.-]*$/;

app.use((_: Request, res: Response, next: NextFunction) => {
    res.setHeader("Content-Disposition", "inline");
    res.setHeader("Content-Type", "text/plain; charset=utf-8");
    next();
});

const decodeDockerLogs = (data: Buffer): string => {
    // ... [no change in this function]
};

app.get("/service/:name/log", (req: Request, res: Response, next: NextFunction) => {
    const containerName: string = req.params.name;

    if (!validNameRegex.test(containerName)) {
        return next(new Error("Invalid container name"));
    }

    const options: RequestOptions = {
        // ... [no change in this part]
    };

    const clientRequest: ClientRequest = request(options, (resp: IncomingMessage) => {
        // ... [no change in this function]
    });

    clientRequest.on("error", (e: Error) => {
        next(e); // Pass the error to the error handling middleware
    });

    clientRequest.end();
});

// Dedicated error handling middleware
app.use((err: Error, req: Request, res: Response, next: NextFunction) => {
    console.error(err.message);
    writeStatusToDisk(statusFilePath, serviceLaunchTime, err.message); // Log the error
    res.status(500).send("An unexpected error occurred. Try again later");
});

app.listen(port, () => {
    console.log(`Logging service listening at http://localhost:${port}`);
});
```

Reply:
What do you think? I do like that it centralises the whole error handling logic

Reply:
So how does it work? `app.use` is handling the errors, and `next` is in charge of surfacing them?